### ISSUE TITLE
Type inference discrepancy: CFE 

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#52813]
Consider the following program:

```dart
class Typer<X> {
  R callWith<R>(R Function<Z>() callback) => callback<X>();
}
 
class A {}
class B extends A {}

void main() {
  Typer<B>().callWith<List>(<X>() => <X>[]); // OK.
  Typer<B>().callWith(<X>() => <X>[]); // Error in analyzer, not CFE.
}
```

This program is accepted with `dart` (commit 63b567add21a9c6a74dae44a13f0588f2fd50e78), but the type inference step fails in the analyzer when no actual type argument is passed to `callback`.

```
Analyzing n024.dart...                 0.9s

  error • n024.dart:13:23 • The argument type 'List<X> Function<X>()' can't
          be assigned to the parameter type 'List<Z0> Function<Z>()'. •
          argument_type_not_assignable

1 issue found.
```

I'm not sure which tool is doing the right thing.

@johnniwinther, @stereotype441, @scheglov, @leafpetersen, WDYT?

**Comment by eernstg** [comment]
Yes, I got that too, using `print(R)`. That would be fine, it is the expected outcome if inference chooses a top type. So the remaining question is: Should the analyzer arrive at the same result (or perhaps some other result, even though that seems very unlikely), or should they both have complained that type inference failed?

**Comment by scheglov** [comment]
@bwilkerson @srawlins For planning purposes. I don't have any special knowledge about type inference in the analyzer.
